### ISSUE TITLE
Compare function under Synchronize Dirs doesn't work on BTRFS (Linux,x86)

### ISSUE THREAD

**Issue by Heavygrass** [doublecmd/doublecmd#2024]
### Version

Double Commander
Version: 1.2.0 alpha
Revision: 11725
Commit: b031cf6
Build date: 2024/12/19
Lazarus: 3.6.0.0
FPC: 3.2.2
Platform: x86_64-Linux-gtk2
OS version: Ubuntu 24.04.1 LTS

Widgetset library: GTK 2.24.33

### Description

Hi!
When clicking Synchronize Dirs -> **Compare** and both dirs are on **BTRFS** FS on Linux - it immediately finishes w/o any error and **shows nothing** in results.
All characters in both paths are English alphabet and sometimes underscore "_", and nothing else. 
Again no kind of any error after clicking Compare.

Same behavior on multiple versions of Kernel 6.8.0.x from Ubuntu and with multiple versions of Doublecmd.
Ubuntu 24.04.1.
Is there's a way to obtain log with Doublecmd or somehow troubleshoot?

Thank you.

**Comment by Heavygrass** [comment]
Still looking for an advice.

**Comment by alexx2000** [comment]
Hi, I cannot reproduce. Works fine for me.

**Comment by Heavygrass** [comment]
> Hi, I cannot reproduce. Works fine for me.

Hi alexx2000,
Whats your setup?
Do you use BTRFS on both sides?

**Comment by Heavygrass** [comment]
I wonder if it could be Cinnamon related. Anybody uses "compare" on Cinnamon?

**Comment by alexx2000** [comment]
> Do you use BTRFS on both sides?

Yes.

> Is there a way to get log for compare?

No. But it uses a same function as file panels to get files. So if file panel display files then SyncDirs should display it too.

**Comment by JonnyTech** [comment]
No problem here too. Maybe your btrfs journal is corrupt. Try clearing it:

`btrfs-zero-log /dev/sdXX`

(replace with your corrent partition identifier)

Any errors reported in `dmesg` and `btrfsck`?

**Comment by Heavygrass** [comment]
> No. But it uses a same function as file panels to get files. 

Whats the name of the function/file that compare uses please?
Is there an instruction how to debug Doublecmd?

> So if file panel display files then SyncDirs should display it too.

May be its a pure interface/Cinnamon issue?

**Comment by Heavygrass** [comment]
@JonnyTech 
> Any errors reported in `btrfsck`?

Not that I can see.
btrfsck output:

`[1/7] checking root items
[2/7] checking extents
^L[3/7] checking free space tree
[4/7] checking fs roots
[5/7] checking only csums items (without verifying data)
[6/7] checking root refs
[7/7] checking quota groups skipped (not enabled on this FS)
found 667255427128 bytes used, no error found
total csum bytes: 647961628
total tree bytes: 3712417792
total fs tree bytes: 2843623424
total extent tree bytes: 173604864
btree space waste bytes: 656058285
file data blocks allocated: 11734141886464
 referenced 1655848992768`

**Comment by JonnyTech** [comment]
> May be its a pure GUI/Cinnamon issue?
That is it just cannot display results?

No, I tested with LMDE6 Cinnamon.

**Comment by JonnyTech** [comment]
> Platform: x86_64-Linux-gtk2

I used the qt build, in case that makes any difference.

**Comment by Heavygrass** [comment]
> I used the qt build, in case that makes any difference.

I guess it does.
How do you get qt build?

**Comment by Heavygrass** [comment]
Can I have both gtk and qt builds installed at the same time?

**Comment by Heavygrass** [comment]
I tried **qt** build - it does WORK on the same path, where gtk does NOT.

Looks like the bug is reproducible only with **gtk** build.

I believe I used [this one](https://sourceforge.net/projects/doublecmd/files/DC%20for%20Linux%2064%20bit/Double%20Commander%201.1.21/doublecmd-1.1.21.qt.x86_64.tar.xz/download)

**Comment by JonnyTech** [comment]
> How do you get qt build?

https://software.opensuse.org//download.html?project=home%3AAlexx2000%3Adoublecmd-svn&package=doublecmd-qt

> Can I have both gtk and qt builds installed at the same time?

Use Portable or AppImage build.

**Comment by Heavygrass** [comment]
@JonnyTech ,
Thanks, but I think you didn't refresh the page.

**Comment by j2969719** [comment]
make sure that your comparison mask is *

both file search and directory synchronization share mask history by default. its possible that you simply had a mask set from the last search

**Comment by Heavygrass** [comment]
> make sure that your comparison mask is *

Thank you, will check.
What is the expected behavior if the mask is **empty**?

**Comment by Heavygrass** [comment]
* mask makes compare to run longer and it shows "Comparing" windows with progress bar, but there's still no results displayed after is completes.
The bottom line shows zero for everything.

**Comment by Heavygrass** [comment]
I get GTK builds from this repository
[http://download.opensuse.org/repositories/home:/Alexx2000:/doublecmd-svn/xUbuntu_24.04/](http://download.opensuse.org/repositories/home:/Alexx2000:/doublecmd-svn/xUbuntu_24.04/)

Could smbd please advise how to switch to QT Ubuntu builds with updates from repository?

**Comment by j2969719** [comment]
>but there's still no results displayed after is completes.

which buttons do you have pressed in the `Show` section? 
https://doublecmd.github.io/doc/en/syncdirs.html#window

>how to switch to QT Ubuntu builds

just install from gui or from terminal with command `sudo apt install doublecmd-qt`, if gtk build was installed from this repo then the package manager will figure out what to do. btw if you installed third-party wlx plugins for gtk build of doublecmd, then you should remove them in the options>plugins>plugins wlx as they are not compatible with qt build and vice versa

happy ny btw

**Comment by Heavygrass** [comment]
> just install from gui or from terminal with command `sudo apt install doublecmd-qt`

I did `sudo apt install doublecmd-qt`, please see output below.
doublecmd runs, but the coloring has **white** background (not like GTK build), and doesn't match Ubuntu default coloring.
GTK build from the same repository had Ubuntu default coloring.
Is there an easy way to make default coloring for QT build?
===========================

`Reading package lists... Done
Building dependency tree... Done
Reading state information... Done
The following packages were automatically installed and are no longer required:
  libebml5 libmatroska7 libpugixml1v5 linux-headers-6.8.0-49 linux-headers-6.8.0-49-generic
  linux-image-6.8.0-49-generic linux-modules-6.8.0-49-generic linux-modules-extra-6.8.0-49-generic
  linux-tools-6.8.0-49 linux-tools-6.8.0-49-generic
Use 'sudo apt autoremove' to remove them.
The following additional packages will be installed:
  doublecmd-common doublecmd-plugins libqt5pas1
Suggested packages:
  doublecmd-help-en | doublecmd-help libffmpegthumbnailer4v5 mplayer rabbitvcs-core
The following packages will be REMOVED:
  doublecmd-gtk
The following NEW packages will be installed:
  doublecmd-qt libqt5pas1
The following packages will be upgraded:
  doublecmd-common doublecmd-plugins
2 upgraded, 2 newly installed, 1 to remove and 2 not upgraded.
Need to get 9,931 kB of archives.
After this operation, 2,636 kB of additional disk space will be used.
Do you want to continue? [Y/n] y
Get:1 http://us.archive.ubuntu.com/ubuntu noble/universe amd64 libqt5pas1 amd64 3.0+dfsg1-1build3 [525 kB]
Get:2 http://download.opensuse.org/repositories/home:/Alexx2000:/doublecmd-svn/xUbuntu_24.04  doublecmd-common 1.2.0~0+svn11766+gitd597ee4-1 [2,760 kB]
Get:3 http://download.opensuse.org/repositories/home:/Alexx2000:/doublecmd-svn/xUbuntu_24.04  doublecmd-plugins 1.2.0~0+svn11766+gitd597ee4-1 [1,315 kB]
Get:4 http://download.opensuse.org/repositories/home:/Alexx2000:/doublecmd-svn/xUbuntu_24.04  doublecmd-qt 1.2.0~0+svn11766+gitd597ee4-1 [5,332 kB]
Fetched 9,931 kB in 4s (2,322 kB/s)       
(Reading database ... 364542 files and directories currently installed.)
Removing doublecmd-gtk (1.2.0~0+svn11760+git112ffaa-1) ...
(Reading database ... 364535 files and directories currently installed.)
Preparing to unpack .../doublecmd-common_1.2.0~0+svn11766+gitd597ee4-1_all.deb ...
Unpacking doublecmd-common (1.2.0~0+svn11766+gitd597ee4-1) over (1.2.0~0+svn11760+git112ffaa-1) ...
Preparing to unpack .../doublecmd-plugins_1.2.0~0+svn11766+gitd597ee4-1_amd64.deb ...
Unpacking doublecmd-plugins (1.2.0~0+svn11766+gitd597ee4-1) over (1.2.0~0+svn11760+git112ffaa-1) ...
Selecting previously unselected package libqt5pas1:amd64.
Preparing to unpack .../libqt5pas1_3.0+dfsg1-1build3_amd64.deb ...
Unpacking libqt5pas1:amd64 (3.0+dfsg1-1build3) ...
Selecting previously unselected package doublecmd-qt.
Preparing to unpack .../doublecmd-qt_1.2.0~0+svn11766+gitd597ee4-1_amd64.deb ...
Unpacking doublecmd-qt (1.2.0~0+svn11766+gitd597ee4-1) ...
Setting up doublecmd-plugins (1.2.0~0+svn11766+gitd597ee4-1) ...
Setting up doublecmd-common (1.2.0~0+svn11766+gitd597ee4-1) ...
Setting up libqt5pas1:amd64 (3.0+dfsg1-1build3) ...
Setting up doublecmd-qt (1.2.0~0+svn11766+gitd597ee4-1) ...
Processing triggers for hicolor-icon-theme (0.17-2) ...
Processing triggers for libc-bin (2.39-0ubuntu8.3) ...
Processing triggers for man-db (2.12.0-4build2) ...
Processing triggers for desktop-file-utils (0.27-2build1) ...`

**Comment by Heavygrass** [comment]
> which buttons do you have pressed in the `Show` section?

In Ubuntu coloring I'm not exactly sure how to determine whether a button is pressed or not, I tried pressing/releasing every button in GTK build.

> plugins wlx as they are not compatible with qt build and vice versa

No wlx plugins.

**Comment by j2969719** [comment]
tldr gtk and qt applications use different theme engines. i havent used ubuntu for ages and do not want to scare you with archwiki. so either look for instructions on installing kvantum in ubuntu (as far as I remember there is a similar theme there to the one used in ubuntu by default) or wait for an answer from more knowledgeable people.

on another note, since both builds use the same config, try to make all the buttons in Show section pressed using the qt build, and then install the gtk build back and check if the situation changes. maybe you just unpressed something earlier and due to the poor readability of the gtk2 theme on ubuntu you simply didnt notice that the buttons were unpressed?

**Comment by Heavygrass** [comment]
Whats the difference between buttons:
"duplicates – Show files that exist on both sides."
and
"= – Show files that are considered identical with the current comparison parameters."?

**Comment by Heavygrass** [comment]
> due to the poor readability

I can distinguish pressed/unpressed colors in GTK build, its rather unclear which color means what.
At least on Cinnamon the **shape of buttons doesn't change** whether  pressed or unpressed, only colors change.
One can only remember what means what.

**Comment by j2969719** [comment]
the duplicates pushbutton here does not mean identical files, but files existing in both folders, if you are only interested in newly added files, by unpressing this button, files existing in both folders will not be displayed 

the equals pushbutton shows files that are considered identical after comparison. they are considered, and not necessarily identical, since you can disable the checkbox by content and the comparison will only be based on basic things like size, name and date (unless the ignore date checkbox is enabled)

**Comment by Heavygrass** [comment]
When I run Doublecmd with `sudo`. Compare does work.

But doesn't work if I run it from **non-root user** with the same exact combination of buttons pressed and filter, and in the same folder.
Both files have read access for everybody (including 'other') and Doublecmd started with the same non-root user can open both files w/o any issue.
Both folders and files belong to the same non-root user.

I observed the described above behavior a number of times by now.

**Comment by JonnyTech** [comment]
> When I run Doublecmd with sudo, Compare does work.

@Heavygrass have you manually edited the DC configuration files?

Close DC and rename the `~/.config/doublecmd/` folder, then run DC again and test compare.

**Comment by Heavygrass** [comment]
@JonnyTech  have you manually edited the DC configuration files?

No, I did not for sure.

**Comment by Heavygrass** [comment]
Changed description and changed it to a relevant feature request that should help to avoid such issues in the future.

**Comment by Heavygrass** [comment]
@alexx2000 @j2969719
Could smbd please remove "bug" & "unable to reproduce" tags?
I changed it to a relevant feature request.
